election-server/src/bin/indexer.rs

```diff
@@ -0,0 +1,570 @@
+            .find_instance_creation(.., *contract_address)
+            .await
+            .context("Could not find contract instance creation block")?;
+        height
+    };
+
+    tracing::info!(
+        "Processing blocks from height {} using node {}",
+        from_height,
+        node_endpoint.uri()
+    );
+
+    let mut blocks_stream = node
+        .get_finalized_blocks_from(from_height)
+        .await
+        .context("Error querying blocks")?;
+    let timeout = std::time::Duration::from_secs(max_behind_s.into());
+    while !stop_flag.load(Ordering::Acquire) {
+        let block = blocks_stream
+            .next_timeout(timeout)
+            .await
+            .with_context(|| format!("Timeout reached for node: {}", node_endpoint.uri()))?;
+        let Some(block) = block else {
+            return Err(anyhow!("Finalized block stream dropped"));
+        };
+        let block_data = process_block(&mut node, block.block_hash, contract_address).await?;
+        if block_sender.send(block_data).await.is_err() {
+            tracing::error!("The database connection has been closed. Terminating node queries.");
+            return Ok(());
+        }
+
+        config.processed_height = Some(block.height);
+    }
+
+    tracing::info!("Service stopped gracefully from exit signal.");
+    Ok(())
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    let config = AppConfig::parse();
+
+    {
+        use tracing_subscriber::prelude::*;
+        let log_filter = tracing_subscriber::filter::Targets::new()
+            .with_target(module_path!(), config.log_level);
+
+        tracing_subscriber::registry()
+            .with(tracing_subscriber::fmt::layer())
+            .with(log_filter)
+            .init();
+    }
+
+    tracing::info!("Service started with configuration: {:?}", config);
+
+    // Since the database connection is managed by the background task we use a
+    // oneshot channel to get the height we should start querying at. First the
+    // background database task is started which then sends the height over this
+    // channel.
+    let (height_sender, height_receiver) = tokio::sync::oneshot::channel();
+    // Create a channel between the task querying the node and the task logging
+    // transactions.
+    let (block_sender, block_receiver) = tokio::sync::mpsc::channel(100);
+    // node/db processes run until the stop flag is triggered.
+    let stop_flag = Arc::new(AtomicBool::new(false));
+    let shutdown_handle = tokio::spawn(set_shutdown(stop_flag.clone()));
+
+    let db_stop = stop_flag.clone();
+    let db_handle = tokio::spawn(async move {
+        let result = run_db_process(
+            config.db_connection,
+            &config.contract_address,
+            block_receiver,
+            height_sender,
+            db_stop,
+        )
+        .await;
+        if let Err(error) = result {
+            tracing::error!("Error happened while running DB process: {:?}", error);
```

Comment:
Don't we want to stop here then?

Reply:
We do stop due to the channel to the DB process being closed, but got no information in the log as to why the DB process failed.

From what I noticed, it made no difference in terms of the service shutting down or not whether the future passed to `tokio::spawn` returned an error or not?

Reply:
I am not sure what you mean. If `db_stop` is set then the run_db_process should return Ok(()).